// react-native-paper-alerts/lib/commonjs/provider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _reactNativePaper = require("react-native-paper");

var _context = _interopRequireDefault(require("./context"));

var _reducer = _interopRequireWildcard(require("./reducer"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const keyboardEvents = _reactNative.Platform.select({
  default: {
    show: 'keyboardDidShow',
    hide: 'keyboardDidHide'
  },
  ios: {
    show: 'keyboardWillShow',
    hide: 'keyboardWillHide'
  }
});

const AlertsProvider = ({
  children,
  options
}) => {
  var _state$options$width, _state$options$width2, _state$options$inputL, _state$options$loginI, _state$options$passwo;

  const [state, dispatch] = (0, _react.useReducer)(_reducer.default, { ..._reducer.initialState,
    ...(options && {
      options
    })
  });
  const [value, setValue] = (0, _react.useState)('');
  const [password, setPassword] = (0, _react.useState)('');
  const dismiss = (0, _react.useCallback)(() => dispatch({
    type: 'DISMISS',
    payload: {}
  }), []);
  const alerts = (0, _react.useMemo)(() => ({
    alert: (title, message, button, alertOptions) => dispatch({
      type: 'ALERT',
      payload: {
        title,
        message,
        button,
        ...(alertOptions && {
          options: alertOptions
        }),
        visible: true
      }
    }),
    prompt: (title, message, callbackOrButtons, type, defaultValue, keyboardType, alertOptions) => {
      setValue(defaultValue !== null && defaultValue !== void 0 ? defaultValue : '');
      setPassword('');

      if (typeof callbackOrButtons === 'function') {
        switch (type) {
          case 'plain-text':
            return dispatch({
              type: 'PROMPT',
              payload: {
                title,
                message,
                keyboardType,
                prompt: [{
                  text: 'Ok',
                  onPress: callbackOrButtons
                }, {
                  text: 'Cancel',
                  style: 'cancel'
                }],
                ...(alertOptions && {
                  options: alertOptions
                }),
                visible: true
              }
            });

          case 'secure-text':
            return dispatch({
              type: 'PROMPT',
              payload: {
                title,
                message,
                keyboardType,
                secureTextEntry: true,
                prompt: [{
                  text: 'Ok',
                  onPress: callbackOrButtons
                }, {
                  text: 'Cancel',
                  style: 'cancel'
                }],
                ...(alertOptions && {
                  options: alertOptions
                }),
                visible: true
              }
            });

          case 'login-password':
            return dispatch({
              type: 'PROMPT',
              payload: {
                title,
                message,
                keyboardType,
                login: [{
                  text: 'Ok',
                  onPress: callbackOrButtons
                }, {
                  text: 'Cancel',
                  style: 'cancel'
                }],
                ...(alertOptions && {
                  options: alertOptions
                }),
                visible: true
              }
            });

          case 'default':
            return dispatch({
              type: 'ALERT',
              payload: {
                title,
                message,
                button: [{
                  text: 'Ok',
                  onPress: callbackOrButtons
                }],
                ...(alertOptions && {
                  options: alertOptions
                }),
                visible: true
              }
            });

          default:
            dispatch({
              type: 'PROMPT',
              payload: {
                title,
                message,
                keyboardType,
                prompt: [{
                  text: 'Ok',
                  onPress: callbackOrButtons
                }, {
                  text: 'Cancel',
                  style: 'cancel'
                }],
                ...(alertOptions && {
                  options: alertOptions
                }),
                visible: true
              }
            });
        }
      } else if (typeof callbackOrButtons === 'undefined') {
        switch (type) {
          case 'plain-text':
            return dispatch({
              type: 'PROMPT',
              payload: {
                title,
                message,
                keyboardType,
                prompt: [{
                  text: 'Ok'
                }, {
                  text: 'Cancel',
                  style: 'cancel'
                }],
                ...(alertOptions && {
                  options: alertOptions
                }),
                visible: true
              }
            });

          case 'secure-text':
            return dispatch({
              type: 'PROMPT',
              payload: {
                title,
                message,
                keyboardType,
                secureTextEntry: true,
                prompt: [{
                  text: 'Ok'
                }, {
                  text: 'Cancel',
                  style: 'cancel'
                }],
                ...(alertOptions && {
                  options: alertOptions
                }),
                visible: true
              }
            });

          case 'login-password':
            return dispatch({
              type: 'PROMPT',
              payload: {
                title,
                message,
                keyboardType,
                login: [{
                  text: 'Ok'
                }, {
                  text: 'Cancel',
                  style: 'cancel'
                }],
                ...(alertOptions && {
                  options: alertOptions
                }),
                visible: true
              }
            });

          case 'default':
            return dispatch({
              type: 'ALERT',
              payload: {
                title,
                message,
                button: [{
                  text: 'Ok'
                }],
                ...(alertOptions && {
                  options: alertOptions
                }),
                visible: true
              }
            });

          default:
            dispatch({
              type: 'PROMPT',
              payload: {
                title,
                message,
                keyboardType,
                prompt: [{
                  text: 'Ok'
                }, {
                  text: 'Cancel',
                  style: 'cancel'
                }],
                ...(alertOptions && {
                  options: alertOptions
                }),
                visible: true
              }
            });
        }
      } else {
        switch (type) {
          case 'plain-text':
            return dispatch({
              type: 'PROMPT',
              payload: {
                title,
                message,
                keyboardType,
                prompt: callbackOrButtons,
                ...(alertOptions && {
                  options: alertOptions
                }),
                visible: true
              }
            });

          case 'secure-text':
            return dispatch({
              type: 'PROMPT',
              payload: {
                title,
                message,
                keyboardType,
                secureTextEntry: true,
                prompt: callbackOrButtons,
                ...(alertOptions && {
                  options: alertOptions
                }),
                visible: true
              }
            });

          case 'login-password':
            return dispatch({
              type: 'PROMPT',
              payload: {
                title,
                message,
                keyboardType,
                login: callbackOrButtons,
                ...(alertOptions && {
                  options: alertOptions
                }),
                visible: true
              }
            });

          case 'default':
            return dispatch({
              type: 'ALERT',
              payload: {
                title,
                message,
                button: callbackOrButtons,
                ...(alertOptions && {
                  options: alertOptions
                }),
                visible: true
              }
            });

          default:
            dispatch({
              type: 'PROMPT',
              payload: {
                title,
                message,
                keyboardType,
                prompt: callbackOrButtons,
                ...(alertOptions && {
                  options: alertOptions
                }),
                visible: true
              }
            });
        }
      }
    }
  }), []);
  const theme = (0, _reactNativePaper.useTheme)();
  const translateY = (0, _react.useRef)(new _reactNative.Animated.Value(0));
  const onKeyboardShown = (0, _react.useCallback)(height => _reactNative.Animated.timing(translateY.current, {
    toValue: -height,
    duration: 250,
    easing: _reactNative.Easing.ease,
    useNativeDriver: true
  }).start(), []);
  const onKeyboardHidden = (0, _react.useCallback)(() => _reactNative.Animated.timing(translateY.current, {
    toValue: 0,
    duration: 250,
    easing: _reactNative.Easing.ease,
    useNativeDriver: true
  }).start(), []);
  (0, _react.useEffect)(() => {
    if (state.visible) {
      _reactNative.Keyboard.addListener(keyboardEvents.show, e => onKeyboardShown(e.endCoordinates.height / 2));

      _reactNative.Keyboard.addListener(keyboardEvents.hide, () => onKeyboardHidden());
    }

    return () => {
      onKeyboardHidden();

      _reactNative.Keyboard.removeAllListeners(keyboardEvents.show);

      _reactNative.Keyboard.removeAllListeners(keyboardEvents.hide);
    };
  }, [onKeyboardHidden, onKeyboardShown, state.visible]);
  return /*#__PURE__*/_react.default.createElement(_context.default.Provider, {
    value: alerts
  }, children, /*#__PURE__*/_react.default.createElement(_reactNativePaper.Portal, null, /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, {
    pointerEvents: "box-none",
    style: [styles.animate, {
      transform: [{
        translateY: translateY.current
      }]
    }]
  }, /*#__PURE__*/_react.default.createElement(_reactNativePaper.Dialog, {
    visible: state.visible,
    dismissable: state.options.cancelable,
    onDismiss: () => {
      if (state.options.cancelable) {
        var _state$options$onDism, _state$options;

        (_state$options$onDism = (_state$options = state.options).onDismiss) === null || _state$options$onDism === void 0 ? void 0 : _state$options$onDism.call(_state$options);
      }

      dismiss();
    },
    style: [styles.web, { ...(state.options.width && {
        width: state.options.width,
        alignSelf: 'center',
        maxWidth: state.options.width
      })
    }, { ...(state.options.maxWidth && {
        maxWidth: (_state$options$width = state.options.width) !== null && _state$options$width !== void 0 ? _state$options$width : state.options.maxWidth
      })
    }, { ...(state.options.minWidth && {
        minWidth: (_state$options$width2 = state.options.width) !== null && _state$options$width2 !== void 0 ? _state$options$width2 : state.options.minWidth
      })
    }]
  }, state.title ? /*#__PURE__*/_react.default.createElement(_reactNativePaper.Dialog.Title, null, state.title) : null, state.message ? /*#__PURE__*/_react.default.createElement(_reactNativePaper.Dialog.Content, null, /*#__PURE__*/_react.default.createElement(_reactNativePaper.Paragraph, null, state.message)) : null, state.prompt ? /*#__PURE__*/_react.default.createElement(_reactNativePaper.Dialog.Content, null, /*#__PURE__*/_react.default.createElement(_reactNativePaper.TextInput, {
    autoFocus: state.options.autoFocus,
    mode: state.options.inputAppearance,
    label: (_state$options$inputL = state.options.inputLabel) !== null && _state$options$inputL !== void 0 ? _state$options$inputL : state.secureTextEntry ? 'Password' : undefined,
    value: value,
    onChangeText: setValue,
    secureTextEntry: state.secureTextEntry,
    keyboardAppearance: state.options.keyboardAppearance,
    keyboardType: state.keyboardType
  })) : state.login ? /*#__PURE__*/_react.default.createElement(_reactNativePaper.Dialog.Content, null, /*#__PURE__*/_react.default.createElement(_reactNativePaper.TextInput, {
    autoFocus: state.options.autoFocus,
    mode: state.options.inputAppearance,
    label: (_state$options$loginI = state.options.loginInputLabel) !== null && _state$options$loginI !== void 0 ? _state$options$loginI : 'Login',
    value: value,
    onChangeText: setValue,
    keyboardAppearance: state.options.keyboardAppearance,
    keyboardType: state.keyboardType,
    style: styles.login
  }), /*#__PURE__*/_react.default.createElement(_reactNativePaper.TextInput, {
    mode: state.options.inputAppearance,
    label: (_state$options$passwo = state.options.passwordInputLabel) !== null && _state$options$passwo !== void 0 ? _state$options$passwo : 'Password',
    value: password,
    onChangeText: setPassword,
    keyboardAppearance: state.options.keyboardAppearance,
    keyboardType: state.keyboardType,
    secureTextEntry: true
  })) : null, state.button ? /*#__PURE__*/_react.default.createElement(_reactNativePaper.Dialog.Actions, {
    style: state.options.stacked && styles.stacked
  }, state.button.map(btn => /*#__PURE__*/_react.default.createElement(_reactNativePaper.Button, {
    key: btn.text,
    onPress: () => {
      var _btn$onPress;

      (_btn$onPress = btn.onPress) === null || _btn$onPress === void 0 ? void 0 : _btn$onPress.call(btn);
      dismiss();
    },
    color: btn.style === 'destructive' ? theme.colors.error : btn.style === 'cancel' ? theme.colors.text : theme.colors.primary,
    uppercase: state.options.uppercase
  }, btn.text))) : state.prompt ? /*#__PURE__*/_react.default.createElement(_reactNativePaper.Dialog.Actions, {
    style: state.options.stacked && styles.stacked
  }, state.prompt.map(btn => /*#__PURE__*/_react.default.createElement(_reactNativePaper.Button, {
    key: btn.text,
    onPress: () => {
      var _btn$onPress2;

      (_btn$onPress2 = btn.onPress) === null || _btn$onPress2 === void 0 ? void 0 : _btn$onPress2.call(btn, value);
      dismiss();
    },
    color: btn.style === 'destructive' ? theme.colors.error : btn.style === 'cancel' ? theme.colors.text : theme.colors.primary,
    uppercase: state.options.uppercase
  }, btn.text))) : state.login ? /*#__PURE__*/_react.default.createElement(_reactNativePaper.Dialog.Actions, {
    style: state.options.stacked && styles.stacked
  }, state.login.map(btn => /*#__PURE__*/_react.default.createElement(_reactNativePaper.Button, {
    key: btn.text,
    onPress: () => {
      var _btn$onPress3;

      (_btn$onPress3 = btn.onPress) === null || _btn$onPress3 === void 0 ? void 0 : _btn$onPress3.call(btn, {
        login: value,
        password
      });
      dismiss();
    },
    color: btn.style === 'destructive' ? theme.colors.error : btn.style === 'cancel' ? theme.colors.text : theme.colors.primary,
    uppercase: state.options.uppercase
  }, btn.text))) : /*#__PURE__*/_react.default.createElement(_reactNativePaper.Dialog.Actions, null, /*#__PURE__*/_react.default.createElement(_reactNativePaper.Button, {
    onPress: dismiss,
    uppercase: state.options.uppercase
  }, "Ok"))))));
};

const styles = _reactNative.StyleSheet.create({
  stacked: {
    flexDirection: 'column',
    alignItems: 'flex-end'
  },
  animate: _reactNative.Platform.OS === 'web' ? {
    flex: 1
  } : {
    flex: 1,
    marginVertical: -300
  },
  login: {
    marginBottom: 16
  },

  /** This is ugly and we should do better */
  web: _reactNative.Platform.OS === 'web' ? {
    maxWidth: 560,
    alignSelf: 'center'
  } : {}
});

var _default = AlertsProvider;
exports.default = _default;
